timemachine/fe/interpolate.py

```diff
@@ -1,217 +1,138 @@
-
-    src_kv = _add_dst_to_src_nonbonded(src_idxs, src_params, dst_idxs)
-    dst_kv = _add_dst_to_src_nonbonded(dst_idxs, dst_params, src_idxs)
-    assert src_kv.keys() == dst_kv.keys()
-
-    res = set()
-
-    for k in src_kv.keys():
-        res.add((tuple(k), src_kv[k], dst_kv[k]))
-
-    return res
-
-
-# special case for torsions
-def _add_dst_to_src_chiral_atom(src_idxs, src_params, dst_idxs):
-    src_kv = dict()
-
-    for idxs, k in zip(src_idxs, src_params):
-        src_kv[idxs] = k
-
-    for idxs in dst_idxs:
-        if idxs not in src_kv:
-            src_kv[idxs] = 0
-
-    return src_kv
-
-
-def align_chiral_atom_idxs_and_params(src_idxs, src_params, dst_idxs, dst_params):
-
-    # sanitize
-    src_idxs = [tuple(p) for p in src_idxs]
-    dst_idxs = [tuple(p) for p in dst_idxs]
-
-    src_kv = _add_dst_to_src_chiral_atom(src_idxs, src_params, dst_idxs)
-    dst_kv = _add_dst_to_src_chiral_atom(dst_idxs, dst_params, src_idxs)
-    assert src_kv.keys() == dst_kv.keys()
-
-    res = set()
-
-    for k in src_kv.keys():
-        res.add((tuple(k), src_kv[k], dst_kv[k]))
-
-    return res
-
-
-def _add_dst_to_src_chiral_bond(src_idxs, src_params, src_signs, dst_idxs, dst_params, dst_signs):
-    src_kv = dict()
-
-    for idxs, k, sign in zip(src_idxs, src_params, src_signs):
-        assert idxs[0] < idxs[-1]
-        key = (idxs, sign)
-        src_kv[key] = k
-
-    for idxs, k, sign in zip(dst_idxs, dst_params, dst_signs):
-        assert idxs[0] < idxs[-1]
-        key = (idxs, sign)
-        if key not in src_kv:
-            src_kv[key] = 0
-
-    return src_kv
+align_nonbonded_idxs_and_params = partial(align_idxs_and_params, make_default=lambda _: (0, 0, 0))
+align_chiral_atom_idxs_and_params = partial(align_idxs_and_params, make_default=lambda _: 0)
+align_torsion_idxs_and_params = partial(
+    align_idxs_and_params,
+    make_default=lambda p: (0, p[1], p[2]),
+    key=lambda idxs, params: (idxs, params[2]),
+    get_idxs=lambda key: key[0],
+)
 
 
 def align_chiral_bond_idxs_and_params(src_idxs, src_params, src_signs, dst_idxs, dst_params, dst_signs):
-
-    # sanitize
-    src_idxs = [tuple(p) for p in src_idxs]
-    dst_idxs = [tuple(p) for p in dst_idxs]
-
-    src_kv = _add_dst_to_src_chiral_bond(src_idxs, src_params, src_signs, dst_idxs, dst_params, dst_signs)
-    dst_kv = _add_dst_to_src_chiral_bond(dst_idxs, dst_params, dst_signs, src_idxs, src_params, src_signs)
-
-    assert src_kv.keys() == dst_kv.keys()
-
-    res = set()
-
-    for key in src_kv.keys():
-        idxs, sign = key
-        src_k = src_kv[key]
-        dst_k = dst_kv[key]
-        res.add((idxs, sign, src_k, dst_k))
-
-    return res
+    return {
+        (idxs, s, p1, p2)
+        for idxs, (s, p1), (_, p2) in align_idxs_and_params(
+            src_idxs,
+            zip(src_signs, src_params),
+            dst_idxs,
+            zip(dst_signs, dst_params),
+            make_default=lambda ps: (ps[0], 0),
```

Comment:
nit: ps -> p for consistency with above
Also add a comment `# p[0] is the sign, 0 is the force constant`

Reply:
Done in 0651472